contributors/design-proposals/node/node-fence.md

```diff
@@ -0,0 +1,528 @@
+
+Depending on the deployment, the fencing controller will have capabilities such as:
+- Power management fencing: powering off or rebooting a node, 
+- Storage fencing: disconnection from specific storage to prevent multiple writers), and
+
+Once the node has been made safe by using one or more of the fencing mechanisms listed above, we can know that either the pods are not running anymore or are not accessing shared resources (storage or network) and we can safely delete their objects from the api server to allow the scheduler to initiate new instance.
+
+The design and implementation acknowledge that other entities, such as the autoscaler, are likely to be present and performing similar monitoring and recovery actions. Therefore it is critical that the fencing controller not create or be susceptible to race conditions.
+
+## User experience
+The loss of a worker node should be transparent to user's of
+StatefulSets.  Recovery time for affected Pods should be bounded and
+short, allowing scale up/down events to proceed as normal afterwards.
+
+In the absence of this feature, an end-user has no ability to safely
+or reliably allow StatefulSets to be recovered and as such end-users
+will not be provided with a mechanism to enable/disable this
+functionality on a set-by-set basis.
+
+### Use Cases
+1. In a bare metal Kubernetes deployment, StatefulSets should not
+   require admin intervention in order to restore capacity when a
+   member Pod was active on a lost worker
+
+1. In a bare metal Kubernetes deployment, StatefulSets should not
+   require admin intervention in order to scale up or down when a
+   member Pod was active on a lost worker
+
+1. In a Cloud Kubernetes deployment, StatefulSets without an
+   autoscaler should not require admin intervention in order to scale
+   up or down when a member Pod was active on a lost worker
+
+In other words, the failure of a worker node should not represent a
+single point of failure for StatefulSets.
+
+## Admin experience
+### Configurations required
+- How to trigger fence devices/apis - general template parameters (e.g. cluster UPS address and credentials) and overrides values per node for specific fields (e.g. ports related to node)
+- How to run a fence flow for each node (each node can be attached to several fence devices\apis and the fence flow can defer)
+
+The following is done by `ConfigMap` objects:
+
+#### Node Fence Config
+For each node that supports fencing we will configure NodeFenceConfig - This object is needed to centralize the information about “how” the node can be “fenced” from the cluster - we will have 3 steps for fence “isolation”, “power-management”, “recovery”. In each we will have a list of methods to perform.
+
+Note: to keep those fence configuration we had two options - Adding attributes in metadata section to the existing node object or create ConfigMap that describe this information for each node with fencing options.
```

Comment:
I see this more as a specific implementation detail of the add-on - I don't consider whatever config defined here as canonical (or an API that other must respect).

Reply:
More as an api that admin must follow to enable fence for node. the controller will validate those maps and report if parameters are missing, and I mention here that it can be also part of the node object which sounds less reasonable to me